src/index.js

```diff
@@ -51,13 +60,26 @@ export default function nodeResolve ( options = {} ) {
 
 			return new Promise( ( fulfil, reject ) => {
 				let disregardResult = false;
+				let packageBrowserField = false;
 
 				resolveId(
 					importee,
 					Object.assign({
 						basedir: dirname( importer ),
 						packageFilter ( pkg ) {
-							if ( useModule && pkg[ 'module' ] ) {
+							if (options.browser && typeof pkg[ 'browser' ] === 'object') {
+								packageBrowserField = Object.keys(pkg[ 'browser' ]).reduce((browser, key) => {
+									browser[ key ] = pkg[ 'browser' ][key];
+									if (key[0] === '.' && !extname(key)) browser[ key + '.js'] = browser[ key + '.json' ] = browser[ key ];
+									return browser;
+								}, {});
+							}
+
+							if (options.browser && typeof pkg[ 'browser' ] === 'string') {
+								pkg[ 'main' ] = pkg[ 'browser' ];
+							} else if (options.browser && pkg[ 'browser' ][ pkg[ 'main' ] ]) {
+								pkg[ 'main' ] = pkg[ 'browser' ][ pkg[ 'main' ] ];
```

Comment:
This is the second key part of the `browser` key logic - which is that if we hit a module, it's effective `main` field is either `browser` (if it is a string) or `browser[pkg.main]` (if it is an object). Luckily we can use packageFilter exactly how we already are to acheive this.